src/prefect/engine/cloud/task_runner.py

```diff
@@ -110,10 +111,37 @@ def call_runner_target_handlers(self, old_state: State, new_state: State) -> Sta
             cloud_state = new_state
             state = self.client.set_task_run_state(
                 task_run_id=task_run_id,
-                version=version,
+                version=version if cloud_state.is_running() else None,
                 state=cloud_state,
                 cache_for=self.task.cache_for,
             )
+        except VersionLockError:
+            state = self.client.get_task_run_state(task_run_id=task_run_id)
+
+            if state.is_running():
+                self.logger.debug(
+                    "Version lock encountered and task {} is already in a running state.".format(
+                        self.task.name
+                    )
+                )
+                raise ENDRUN(state=state)
+
+            self.logger.debug(
+                "Version lock encountered for task {}, proceeding with state {}...".format(
+                    self.task.name, type(state).__name__
+                )
+            )
+
+            try:
+                new_state = state.load_result(self.result)
```

Comment:
I'm curious - should this ENDRUN rather than return the new state, to abort further processing of this task? For example, let's say it tries to enter a RUNNING state, but learns that the "true" state is Failed. I don't think we want to enter the RETRYING part of this task runner's pipeline, which we might if we just return the upstream state. Therefore I'm wondering if anytime we learn that the true state is different from what we thought, we just ENDRUN with it and let the next task proceed.

Not 100% sure if I'm thinking through everything here - what do you think?